kms/capi/capi.go

```diff
@@ -0,0 +1,715 @@
+		storeLocation = "user"
+	}
+
+	certStoreLocation := certStoreCurrentUser
+	switch storeLocation {
+	case "user":
+		certStoreLocation = certStoreCurrentUser
+	case "machine":
+		certStoreLocation = certStoreLocalMachine
+	default:
+		return nil, fmt.Errorf("invalid cert store location %v", storeLocation)
+	}
+
+	var storeName string
+
+	// default to the 'My' store
+	if storeName = u.Get(StoreNameArg); storeName == "" {
+		storeName = "My"
+	}
+
+	st, err := windows.CertOpenStore(
+		certStoreProvSystem,
+		0,
+		0,
+		certStoreLocation,
+		uintptr(unsafe.Pointer(wide(storeName))))
+	if err != nil {
+		return nil, fmt.Errorf("CertOpenStore for the %v store %v returned: %w", storeLocation, storeName, err)
+	}
+
+	var certHandle *windows.CertContext
+
+	if sha1Hash != "" {
+		sha1Bytes, err := hex.DecodeString(sha1Hash)
+		if err != nil {
+			return nil, fmt.Errorf("%s must be in hex format: %w", HashArg, err)
+		}
+
+		searchData := CERT_ID_KEYIDORHASH{
+			idChoice: CERT_ID_SHA1_HASH,
+			KeyIDOrHash: CRYPTOAPI_BLOB{
+				len:  uint32(len(sha1Bytes)),
+				data: uintptr(unsafe.Pointer(&sha1Bytes[0])),
+			},
+		}
+
+		certHandle, err = findCertificateInStore(st,
+			encodingX509ASN|encodingPKCS7,
+			0,
+			findCertID,
+			uintptr(unsafe.Pointer(&searchData)), nil)
+
+		if err != nil {
+			return nil, fmt.Errorf("findCertificateInStore failed: %w", err)
+		}
+
+		if certHandle == nil {
+			return nil, fmt.Errorf("certificate with %v=%s not found", KeyIDArg, keyID)
+		}
+
+		defer windows.CertFreeCertificateContext(certHandle)
+		return certContextToX509(certHandle)
+	} else if keyID != "" {
+		keyIDBytes, err := hex.DecodeString(keyID)
+		if err != nil {
+			return nil, fmt.Errorf("%v must be in hex format: %w", KeyIDArg, err)
+		}
+
+		searchData := CERT_ID_KEYIDORHASH{
+			idChoice: CERT_ID_KEY_IDENTIFIER,
+			KeyIDOrHash: CRYPTOAPI_BLOB{
+				len:  uint32(len(keyIDBytes)),
+				data: uintptr(unsafe.Pointer(&keyIDBytes[0])),
+			},
+		}
+
+		certHandle, err = findCertificateInStore(st,
+			encodingX509ASN|encodingPKCS7,
+			0,
+			findCertID,
+			uintptr(unsafe.Pointer(&searchData)), nil)
+
+		if err != nil {
+			return nil, fmt.Errorf("findCertificateInStore failed: %w", err)
+		}
+
+		if certHandle == nil {
+			return nil, fmt.Errorf("certificate with %v=%s not found", KeyIDArg, keyID)
+		}
+
+		defer windows.CertFreeCertificateContext(certHandle)
+		return certContextToX509(certHandle)
+	} else if issuerName != "" && serialNumber != "" {
+		serialBytes, err := hex.DecodeString(serialNumber)
+		if err != nil {
+			return nil, fmt.Errorf("%v must be in hex format: %w", SerialNumberArg, err)
+		}
+
+		// iterate over all certificates from issuer, and check the SN
+		var prevCert *windows.CertContext = nil
+		for {
+			certHandle, err = findCertificateInStore(st,
+				encodingX509ASN|encodingPKCS7,
+				0,
+				findIssuerStr,
+				uintptr(unsafe.Pointer(wide(issuerName))), prevCert)
```

Comment:
Can't we use [CERT_ISSUER_SERIAL_NUMBER](https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/ns-wincrypt-cert_issuer_serial_number) instead? It is one of the members of the [CERT_ID](https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/ns-wincrypt-cert_id) union that we use above. It will make the for loop unnecessary, as according to the docs "The combination of the issuer name and the serial number is a unique identifier of a certificate."

Do you know what is exactly the "issuer name"? The full issuer `C=US,ST=California,L=San Francisco,O=GitHub, Inc.,CN=github.com`? The CommonName?

Reply:
I would love to be able to use [CERT_ISSUER_SERIAL_NUMBER](https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/ns-wincrypt-cert_issuer_serial_number), but i spent quite a few hours not being able to. I believe there is something weird where that search criteria needs absolutely exact matches, but CERT_FIND_ISSUER_STR does something else which allows fuzzier matches?

So to make it work i used the combination for now, its sub-optimal but I couldnt find any resources or examples on it to guide me and I wasnt making any progress with it unfortunately.

the "issuer name" is the full `pkix.Name`

Reply:
I haven't tested this, but this Russian forum might provide some insight
https://www.cryptopro.ru/forum2/default.aspx?g=posts&t=15614